### PR TITLE
Add spec compliant test suite

### PR DESCRIPTION
**Pull request by benjiro**
Implements a spec-compliant test suite https://github.com/open-feature/dotnet-sdk/issues/14
- Fix singleton so it is implemented as per the design pattern
- Add strong typed Error Code

### CODE REVIEW

**Comment by beeme1mr** [comment]
@toddbaert, can you please take a look at this tomorrow?

**Comment by toddbaert** [comment]
Looking at this now.

File: src/OpenFeature/Model/ResolutionDetails.cs [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
     {
         public T Value { get; }
         public string FlagKey { get; }
-        public string ErrorCode { get; }
+        public ErrorType ErrorType { get; }
         public string Reason { get; }
         public string Variant { get; }
 
-        public ResolutionDetails(string flagKey, T value, string errorCode = null, string reason = null, string variant = null)
+        public ResolutionDetails(string flagKey, T value, ErrorType errorType = ErrorType.None, string reason = null, string variant = null)
```
Comment (toddbaert) [inline src/OpenFeature/Model/ResolutionDetails.cs:13]: I spent more time than I'd like to admit thinking about the parameter order here given the optionals.. I guess it doesn't matter to much, but I think this order and these defaults makes the most sense.

File: src/OpenFeature/OpenFeatureClient.cs [hunk 2]
```diff
@@ -80,76 +85,69 @@ public FeatureClient(IFeatureProvider featureProvider, string name, string versi
...
-                    {
-                        context.EvaluationContext.Merge(resp);
-                    }
-                    else
-                    {
-                        OpenFeature.Logger.LogDebug("Hook {HookName} returned null, nothing to merge back into context", hook.GetType().Name);
-                    }
+                    context.EvaluationContext.Merge(resp);
                 }
-                catch (Exception e)
```
Comment (toddbaert) [inline src/OpenFeature/OpenFeatureClient.cs:130]: So before errors were being swallowed, now they are thrown and handled in the evaluation, which makes sense. :+1:

File: test/OpenFeature.Tests/OpenFeatureClientTests.cs [hunk 3]
```diff
@@ -3,16 +3,182 @@
...
+        {
+            var fixture = new Fixture();
+            var clientName = fixture.Create<string>();
+            var clientVersion = fixture.Create<string>();
+            var flagName = fixture.Create<string>();
+            var defaultValue = fixture.Create<TestStructure>();
+            var mockedFeatureProvider = new Mock<IFeatureProvider>();
+            var mockedLogger = new Mock<ILogger<OpenFeature>>();
+
+            // This will fail to case a String to TestStructure
```
Comment (toddbaert) [inline test/OpenFeature.Tests/OpenFeatureClientTests.cs:154]: I'm not sure if I understand this comment :sweat_smile:

File: test/OpenFeature.Tests/OpenFeatureHookTests.cs [hunk 4]
```diff
@@ -1,111 +1,371 @@
...
         {
-            var fixture = new Fixture();
-            var clientName = fixture.Create<string>();
-            var clientVersion = fixture.Create<string>();
-            var flagName = fixture.Create<string>();
-            var defaultValue = fixture.Create<bool>();
+            var featureProvider = new Mock<IFeatureProvider>();
+            var hook = new Mock<Hook>();
+
+            var sequence = new MockSequence();
```
Comment (toddbaert) [inline test/OpenFeature.Tests/OpenFeatureHookTests.cs:188]: This `MockSequence` API is very cool.

**Comment by toddbaert** [comment]
I tried to find any significant behavior that was uncovered, and I cannot. This is great, the tests are very readable to me (I'd never seen the `InlineData` decorator before, that's a nice API) and the spec numbering also helps.

Thanks again!

**Review by toddbaert — APPROVED** [review]

**Pull request merged**